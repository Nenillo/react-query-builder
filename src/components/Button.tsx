import React from 'react';
import styled from 'styled-components';
import { colors } from '../constants/colors';

const StyledButton = styled.button`
  color: #fff;
  border: 0;
  background-color: ${colors.primary};
  padding: 0.5rem 1rem;
  font-size: 0.7rem;
  white-space: nowrap;
  cursor: pointer;
  border-radius: 3px;
  outline: none;
`;

export interface ButtonProps {
  onClick: () => void;
  className?: string;
  label: string;
}

export const Button: React.FC<ButtonProps> = ({
  onClick,
  className,
  label,
}) => (
  <StyledButton onClick={onClick} className={className}>
    {label}
  </StyledButton>
);
