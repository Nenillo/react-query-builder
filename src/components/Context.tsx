import React, { createContext } from 'react';
import {
  BuilderFieldProps,
  BuilderComponentsProps,
  defaultComponents,
} from './Builder';
import { strings as defaultStrings, Strings } from '../constants/strings';

export interface BuilderContextProps {
  fields: BuilderFieldProps[];
  data: any;
  components: BuilderComponentsProps;
  strings: Strings;
  setData: React.Dispatch<any>;
  onChange: (data: any) => void;
}

export const BuilderContext = createContext<BuilderContextProps>(
  {} as BuilderContextProps
);

export interface BuilderContextProviderProps extends BuilderContextProps {
  children: React.ReactNode | React.ReactNodeArray;
}

export const BuilderContextProvider: React.FC<BuilderContextProviderProps> = ({
  fields,
  components,
  strings,
  data,
  setData,
  onChange,
  children,
}) => {
  components = {
    ...defaultComponents,
    ...components,
    form: { ...defaultComponents.form, ...components.form },
  };

  strings = {
    ...defaultStrings,
    ...strings,
    component: {
      ...defaultStrings.component,
      ...strings.component,
    },
    form: {
      ...defaultStrings.form,
      ...strings.form,
    },
    group: {
      ...defaultStrings.group,
      ...strings.group,
    },
    operators: {
      ...defaultStrings.operators,
      ...strings.operators,
    },
  };

  return (
    <BuilderContext.Provider
      value={{
        fields,
        components,
        strings,
        data,
        setData,
        onChange,
      }}
    >
      {children}
    </BuilderContext.Provider>
  );
};
