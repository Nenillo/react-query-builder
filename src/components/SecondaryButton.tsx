import { Button, ButtonProps } from './Button';
import styled from 'styled-components';
import { colors } from '../constants/colors';

export const SecondaryButton = styled(Button)<ButtonProps>`
  background-color: ${colors.tertiary};
`;
