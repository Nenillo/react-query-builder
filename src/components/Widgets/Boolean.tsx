import React, { useContext } from 'react';
import { clone } from '../../utils/clone';
import { BuilderContext } from '../Context';

export interface BooleanProps {
  selectedValue: boolean;
  id: string;
}

// tslint:disable-next-line: variable-name
export const Boolean: React.FC<BooleanProps> = ({ selectedValue, id }) => {
  const { data, setData, onChange, components, readOnly } = useContext(
    BuilderContext
  );

  const { form } = components;

  const handleChange = (value: boolean) => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);

    clonedData[parentIndex].value = value;

    setData(clonedData);
    onChange(clonedData);
  };

  if (form) {
    return (
      <form.Switch
        onChange={handleChange}
        switched={selectedValue}
        disabled={readOnly}
      />
    );
  }

  return null;
};
