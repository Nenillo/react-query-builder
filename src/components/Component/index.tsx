import React, { useContext } from 'react';
import styled from 'styled-components';
import { BuilderFieldOperator, BuilderContext } from '../Builder';
import { FieldSelect } from '../Widgets/FieldSelect';
import { Boolean } from '../Widgets/Boolean';
import { Select } from '../Widgets/Select';
import { SelectMulti } from '../Widgets/SelectMulti';
import { OperatorSelect } from '../Widgets/OperatorSelect';
import { Input } from '../Widgets/Input';
import {
  isBoolean,
  isReactText,
  isReactTextArray,
  isOptionList,
} from '../../utils/types';
import { Component as ComponentContainer } from './Component';

const BooleanContainer = styled.div`
  align-self: center;
`;

export interface ComponentProps {
  field: string;
  value?: React.ReactText | React.ReactText[] | boolean;
  operator?: BuilderFieldOperator;
  id: string;
}

export const Component: React.FC<ComponentProps> = ({
  field: fieldRef,
  value: selectedValue,
  operator,
  id,
}) => {
  const { fields } = useContext(BuilderContext);

  if (fields) {
    if (fieldRef === '') {
      return (
        <ComponentContainer>
          <FieldSelect selectedValue={''} id={id} />
        </ComponentContainer>
      );
    } else {
      try {
        const fieldIndex = fields.findIndex(item => item.field === fieldRef);
        const { field, operators, type, value: fieldValue } = fields[
          fieldIndex
        ];

        const operatorsOptionList =
          operators && operators.map(item => ({ value: item, label: item }));

        return (
          <ComponentContainer>
            <FieldSelect selectedValue={field} id={id} />

            {type === 'BOOLEAN' && isBoolean(selectedValue) && (
              <BooleanContainer>
                <Boolean id={id} selectedValue={selectedValue} />
              </BooleanContainer>
            )}

            {type === 'LIST' &&
              isReactText(selectedValue) &&
              isOptionList(fieldValue) &&
              isOptionList(operatorsOptionList) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Select
                    id={id}
                    selectedValue={selectedValue}
                    values={fieldValue}
                  />
                </>
              )}

            {type === 'MULTI_LIST' &&
              isOptionList(fieldValue) &&
              isOptionList(operatorsOptionList) &&
              isReactTextArray(selectedValue) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <SelectMulti
                    id={id}
                    values={fieldValue}
                    selectedValue={selectedValue}
                  />
                </>
              )}

            {type === 'TEXT' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="text" value={selectedValue} id={id} />
                </>
              )}

            {type === 'NUMBER' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="number" value={selectedValue} id={id} />
                </>
              )}

            {type === 'DATE' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="date" value={selectedValue} id={id} />
                </>
              )}
          </ComponentContainer>
        );
      } catch (e) {
        console.error(`Field "${fieldRef}" not found in fields definition.`);
      }
    }
  }

  return null;
};
