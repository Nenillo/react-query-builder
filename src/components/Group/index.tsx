import React, { useContext } from 'react';
import { BuilderGroupValues, BuilderContext } from '../Builder';
import uniqid from 'uniqid';
import { clone } from '../../utils/clone';

export interface GroupProps {
  value?: BuilderGroupValues;
  isNegated?: boolean;
  children?: React.ReactNode | React.ReactNodeArray;
  id: string;
}

export const Group: React.FC<GroupProps> = ({
  value,
  isNegated,
  children,
  id,
}) => {
  const { components, data, setData, onChange } = useContext(BuilderContext);

  const { Add, Group: GroupContainer, GroupHeaderOption: Option } = components;

  const findIndex = () => {
    const clonedData = clone(data);
    const parentIndex = clonedData.findIndex((item: any) => item.id === id);
    let insertAfter = parentIndex;

    if (data[parentIndex].children) {
      const lastChildren = clonedData[parentIndex].children.slice(-1)[0];
      insertAfter = clonedData.findIndex(
        (item: any) => item.id === lastChildren
      );
    }

    return { insertAfter, parentIndex, clonedData };
  };

  const addItem = (payload: any) => {
    const { insertAfter, parentIndex, clonedData } = findIndex();

    if (!clonedData[parentIndex].children) {
      clonedData[insertAfter].children = [];
    }

    clonedData[parentIndex].children.push(payload.id);
    clonedData.splice(insertAfter, 0, payload);

    setData(clonedData);
    onChange(clonedData);
  };

  const handleAddGroup = () => {
    const EmptyGroup: any = {
      type: 'GROUP',
      value: 'AND',
      isNegated: false,
      id: uniqid(),
      parent: id,
      children: [],
    };

    addItem(EmptyGroup);
  };

  const handleAddRule = () => {
    const EmptyRule: any = {
      field: '',
      id: uniqid(),
      parent: id,
    };

    addItem(EmptyRule);
  };

  const handleChangeGroupType = (nextValue: BuilderGroupValues) => {
    const { clonedData, parentIndex } = findIndex();
    clonedData[parentIndex].value = nextValue;

    setData(clonedData);
    onChange(clonedData);
  };

  const handleToggleNegateGroup = (nextValue: boolean) => {
    const { clonedData, parentIndex } = findIndex();
    clonedData[parentIndex].isNegated = nextValue;

    setData(clonedData);
    onChange(clonedData);
  };

  return (
    <GroupContainer
      controlsLeft={
        <>
          <Option
            isSelected={!!isNegated}
            value={!isNegated}
            onClick={handleToggleNegateGroup}
          >
            NOT
          </Option>
          <Option
            isSelected={value === 'AND'}
            value="AND"
            onClick={handleChangeGroupType}
          >
            AND
          </Option>
          <Option
            isSelected={value === 'OR'}
            value="OR"
            onClick={handleChangeGroupType}
          >
            OR
          </Option>
        </>
      }
      controlsRight={
        <>
          <Add onClick={handleAddRule} label="Add Rule" />
          <Add onClick={handleAddGroup} label="Add Group" />
        </>
      }
    >
      {children}
    </GroupContainer>
  );
};
