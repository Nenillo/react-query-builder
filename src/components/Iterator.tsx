import React from 'react';
import { Component, ComponentProps } from './Component/index';
import { Group, GroupProps } from './Group/index';

export interface IteratorProps {
  originalData: any;
  filteredData: any;
}

export const Iterator: React.FC<IteratorProps> = ({
  originalData,
  filteredData,
}) => (
  <>
    {filteredData.map((item: any) => {
      if (typeof item.children !== 'undefined') {
        const items: any = [];

        item.children.forEach((id: any) => {
          items.push(originalData.filter((fitem: any) => id === fitem.id)[0]);
        });

        switch (item.type) {
          case 'GROUP': {
            const { id, value, isNegated } = item as GroupProps;

            return (
              <Group key={id} value={value} isNegated={isNegated} id={id}>
                <Iterator originalData={originalData} filteredData={items} />
              </Group>
            );
          }
          default: {
            const { field, value, id, operator } = item as ComponentProps;

            return (
              <div key={item.id}>
                <Component
                  key={id}
                  field={field}
                  value={value}
                  operator={operator}
                  id={id}
                />
                <Iterator originalData={originalData} filteredData={items} />
              </div>
            );
          }
        }
      } else {
        const { field, value, id, operator } = item as ComponentProps;

        return (
          <Component
            key={id}
            field={field}
            value={value}
            operator={operator}
            id={id}
          />
        );
      }
    })}
  </>
);
