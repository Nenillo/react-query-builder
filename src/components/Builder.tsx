import React, { createContext, useState, useEffect } from 'react';
import styled from 'styled-components';
import { Iterator } from './Iterator';
import { colors } from '../constants/colors';
import { normalizeTree } from '../utils/normalizeTree';
import { denormalizeTree } from '../utils/denormalizeTree';
import uniqid from 'uniqid';
import { assignIds } from '../utils/assignIds';

/* Configurable components */

import { Button } from './Button';
import { SecondaryButton } from './SecondaryButton';
import { Input } from './Form/Input';
import { Select } from './Form/Select';
import { SelectMulti } from './Form/SelectMulti';
import { Switch } from './Form/Switch';
import { Component } from './Component/Component';
import { Group } from './Group/Group';
import { Option as GroupHeaderOption } from './Group/Option';

/* Strings */

import { strings as defaultStrings, Strings } from '../constants/strings';

export const StyledBuilder = styled.div`
  padding: 1rem;
  border: 1px solid ${colors.light};
  background: #fff;
`;

export type BuilderFieldType =
  | 'BOOLEAN'
  | 'TEXT'
  | 'DATE'
  | 'NUMBER'
  | 'STATEMENT'
  | 'LIST'
  | 'MULTI_LIST'
  | 'GROUP';

export type BuilderFieldOperator =
  | 'LARGER'
  | 'SMALLER'
  | 'LARGER_EQUAL'
  | 'SMALLER_EQUAL'
  | 'EQUAL'
  | 'NOT_EQUAL'
  | 'ALL_IN'
  | 'ANY_IN'
  | 'NOT_IN'
  | 'BETWEEN'
  | 'NOT_BETWEEN';

export type BuilderGroupValues = 'AND' | 'OR';

export interface BuilderFieldProps {
  field: string;
  label: string;
  value?:
    | string
    | string[]
    | boolean
    | { value: React.ReactText; label: string }[];
  type: BuilderFieldType;
  /* List of available operators */
  operators?: BuilderFieldOperator[];
}

export interface BuilderComponentsProps {
  form?: {
    Select?: any;
    SelectMulti?: any;
    Switch?: any;
    Input?: any;
  };
  Remove?: any;
  Add?: any;
  Component?: any;
  Group?: any;
  GroupHeaderOption?: any;
}

export interface BuilderProps {
  fields: BuilderFieldProps[];
  data: any;
  components?: BuilderComponentsProps;
  strings?: Strings;
  onChange?: (data: any) => any;
}

export interface BuilderContextProps {
  fields: BuilderFieldProps[];
  data: any;
  components: BuilderComponentsProps;
  strings: Strings;
  setData: React.Dispatch<any>;
  onChange: (data: any) => void;
}

export const BuilderContext = createContext<BuilderContextProps>(
  {} as BuilderContextProps
);

export const defaultComponents: BuilderComponentsProps = {
  form: {
    Input,
    Select,
    SelectMulti,
    Switch,
  },
  Remove: SecondaryButton,
  Add: Button,
  Component,
  Group,
  GroupHeaderOption,
};

export const Builder: React.FC<BuilderProps> = ({
  data: originalData,
  fields,
  components = defaultComponents,
  strings = defaultStrings,
  onChange,
}) => {
  let normalizedData: any;
  originalData = assignIds(originalData);

  components = {
    ...defaultComponents,
    ...components,
    form: { ...defaultComponents.form, ...components.form },
  };

  strings = {
    ...defaultStrings,
    ...strings,
    component: {
      ...defaultStrings.component,
      ...strings.component,
    },
    form: {
      ...defaultStrings.form,
      ...strings.form,
    },
    group: {
      ...defaultStrings.group,
      ...strings.group,
    },
    operators: {
      ...defaultStrings.operators,
      ...strings.operators,
    },
  };

  if (originalData.length === 0) {
    originalData = [
      {
        type: 'GROUP',
        value: 'AND',
        isNegated: false,
        id: uniqid(),
        children: originalData,
      },
    ];
  }

  try {
    normalizedData = normalizeTree(originalData);
  } catch (e) {
    throw new Error('Input data tree is in invalid format');
  }

  const [data, setData] = useState(normalizedData);
  const filteredData = data.filter((item: any) => !item.parent);

  useEffect(() => {
    handleChange(normalizedData);
  }, []);

  const handleChange = (nextData: any) => {
    if (onChange) {
      try {
        onChange(denormalizeTree(nextData));
      } catch (e) {
        throw new Error('Input data tree is in invalid format');
      }
    }
  };

  return (
    <BuilderContext.Provider
      value={{
        fields,
        components,
        strings,
        data,
        setData,
        onChange: handleChange,
      }}
    >
      <StyledBuilder>
        <Iterator originalData={data} filteredData={filteredData} />
      </StyledBuilder>
    </BuilderContext.Provider>
  );
};
