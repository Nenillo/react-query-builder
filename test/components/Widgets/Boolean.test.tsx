import { mount, shallow } from 'enzyme';
import React from 'react';
import {
  BuilderComponentsProps,
  BuilderFieldProps,
  defaultComponents,
} from '../../../src/components/Builder';
import { BuilderContext } from '../../../src/components/Context';
import { Boolean } from '../../../src/components/Widgets/Boolean';

const components: BuilderComponentsProps = defaultComponents;
const fields: BuilderFieldProps[] = [
  {
    field: 'MOCK_FIELD',
    label: 'Mock Field',
    type: 'BOOLEAN',
  },
];
const data: any[] = [
  {
    id: 'test',
    value: false,
  },
];
const strings = {};
const setData = jest.fn();
const onChange = () => jest.fn();

describe('#components/Widgets/Boolean', () => {
  it('Tests Snapshot', () => {
    expect(
      shallow(
        <BuilderContext.Provider
          value={{
            components,
            fields,
            data,
            strings,
            setData,
            onChange,
            readOnly: false,
          }}
        >
          <Boolean id="test" selectedValue={false} />
        </BuilderContext.Provider>
      )
    ).toMatchSnapshot();

    expect(
      shallow(
        <BuilderContext.Provider
          value={{
            components,
            fields,
            data,
            strings,
            setData,
            onChange,
            readOnly: true,
          }}
        >
          <Boolean id="test" selectedValue={false} />
        </BuilderContext.Provider>
      )
    ).toMatchSnapshot();
  });

  it('Tests user interaction', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{
          components,
          fields,
          data,
          strings,
          setData,
          onChange,
          readOnly: false,
        }}
      >
        <Boolean id="test" selectedValue={false} />
      </BuilderContext.Provider>
    );

    const switchEl = wrapper.find('[data-test="Switch"]').first();
    switchEl.simulate('click');
  });

  it('Tests no form components scenario', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{
          components: {},
          fields,
          data,
          strings,
          setData,
          onChange,
          readOnly: false,
        }}
      >
        <Boolean id="test" selectedValue={false} />
      </BuilderContext.Provider>
    );

    expect(Object.keys(wrapper).length).toBe(0);
  });
});
