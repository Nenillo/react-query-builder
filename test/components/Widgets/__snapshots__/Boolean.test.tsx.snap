// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#components/Widgets/Boolean Tests Snapshot 1`] = `
<Component
  id="test"
  selectedValue={false}
/>
`;

exports[`#components/Widgets/Boolean Tests Snapshot 2`] = `
<Component
  id="test"
  selectedValue={false}
/>
`;
